plan/eliminate_projection.go

```diff
@@ -14,50 +14,229 @@
+}
+
+// eliminatePhysicalProjection should be called after physical optimization to eliminate the redundant projection
+// left after logical projection elimination.
+func eliminatePhysicalProjection(p PhysicalPlan, replace map[string]*expression.Column) PhysicalPlan {
+	children := make([]Plan, 0, len(p.Children()))
+	for _, child := range p.Children() {
+		children = append(children, eliminatePhysicalProjection(child.(PhysicalPlan), replace))
+	}
+	p.SetChildren(children...)
+
+	for _, dst := range p.Schema().Columns {
+		resolveColumnAndReplace(dst, replace)
+	}
+	for _, key := range p.Schema().Keys {
+		for _, keyCol := range key {
+			resolveColumnAndReplace(keyCol, replace)
+		}
+	}
+	p.replaceExprColumns(replace)
+
+	proj, isProj := p.(*Projection)
+	if !isProj || !canProjectionBeEliminatedStrict(proj) {
+		return p
+	}
+	child := p.Children()[0]
+	exprs := proj.Exprs
+	childCols := child.Schema().Columns
+	for i, parentColumn := range proj.Schema().Columns {
+		col, _ := exprs[i].(*expression.Column)
+		col.DBName = parentColumn.DBName
+		col.TblName = parentColumn.TblName
+		col.ColName = parentColumn.ColName
+		childCols[i].ColName = parentColumn.ColName
+		replace[string(parentColumn.HashCode())] = col
+	}
+	RemovePlan(p)
+	return child.(PhysicalPlan)
+}
+
+type projectionEliminater struct {
+}
+
+// optimize implements the logicalOptRule interface
+func (pe *projectionEliminater) optimize(lp LogicalPlan, _ context.Context, _ *idAllocator) (LogicalPlan, error) {
+	root := pe.eliminate(lp, make(map[string]*expression.Column), false)
+	return root.(LogicalPlan), nil
+}
+
+// eliminate eliminates the redundant projection in a logical plan.
+// The order of output plan's schema columns can be changed if "haveProjection" is true.
+func (pe *projectionEliminater) eliminate(p LogicalPlan, replace map[string]*expression.Column, haveProjection bool) LogicalPlan {
+	proj, isProj := p.(*Projection)
+	children := make([]Plan, 0, len(p.Children()))
+	_, isUnion := p.(*Union)
+	for _, child := range p.Children() {
+		children = append(children, pe.eliminate(child.(LogicalPlan), replace, !isUnion && (haveProjection || isProj)))
+	}
+	p.SetChildren(children...)
+
+	switch p.(type) {
+	case *Sort, *TopN, *Limit, *Selection, *MaxOneRow, *Update, *SelectLock:
+		p.SetSchema(p.Children()[0].Schema())
+	case *LogicalJoin, *LogicalApply:
+		var joinTp JoinType
+		if _, isApply := p.(*LogicalApply); isApply {
+			joinTp = p.(*LogicalApply).JoinType
+		} else {
+			joinTp = p.(*LogicalJoin).JoinType
+		}
+		switch joinTp {
+		case InnerJoin, LeftOuterJoin, RightOuterJoin:
+			p.SetSchema(expression.MergeSchema(p.Children()[0].Schema(), p.Children()[1].Schema()))
+		case SemiJoin:
+			p.SetSchema(p.Children()[0].Schema().Clone())
+		case LeftOuterSemiJoin:
+			newSchema := p.Children()[0].Schema().Clone()
+			newSchema.Append(p.Schema().Columns[len(p.Schema().Columns)-1])
+			p.SetSchema(newSchema)
+		}
+	default:
+		for _, dst := range p.Schema().Columns {
```

Comment:
Aggregation function "firstrow"'s corresponding output column is cloned from its child's. the `default` branch has to be retained.